wurst/systems/spawns/ResourceSpawnsUtils.wurst

```diff
@@ -0,0 +1,197 @@
+package ResourceSpawnsUtils
+
+import Composer
+
+// Standard library imports:
+import HashMap
+import Interpolation
+import TerrainUtils
+
+// Local imports:
+import Game
+import GameTimer
+import GameConfig
+import GeometryUtils
+import Pool
+
+function getRandomAllowedPoint(Pool<rect> allowedRegions, region restrainedSpawnRegion) returns vec2
+    let spawnRect = allowedRegions.random()
+
+    while true
+        let spawnPoint = spawnRect.randomPoint()
+        if spawnPoint.isTerrainLand() and spawnPoint.isTerrainWalkable() and not spawnPoint.isInRegion(restrainedSpawnRegion)
+            return spawnPoint
+    
+    return vec2(0,0)
+
+public class Spawner
+    Composer fn
+    
+    construct(Composer fn)
+        this.fn = fn
+    
+    function spawn()
+        this.fn.execute()
+
+public class ItemSpawnComposer implements Composer
+    protected ItemPool items
+    protected Pool<rect> allowedRegions
+    protected region disallowedRegions
+
+    construct(ItemPool items, Pool<rect> allowedRegions, region disallowedRegions)
+        this.items = items
+        this.allowedRegions = allowedRegions
+        this.disallowedRegions = disallowedRegions
+    
+    override function execute()
+        let spawnPoint = getRandomAllowedPoint(this.allowedRegions, this.disallowedRegions)
+        let spawnItemId = this.items.random()
+        createItem(spawnItemId, spawnPoint)
+
+        if not this.items.spawnsPerType.has(spawnItemId)
+            this.items.spawnsPerType.put(spawnItemId, 1)
+        else
+            this.items.spawnsPerType.put(spawnItemId, this.items.spawnsPerType.get(spawnItemId) + 1)
+
+public class FishSpawnComposer implements Composer
+    protected Pool<int> units
+    protected rect allowedRegion
+
+    construct(Pool<int> units, rect allowedRegion)
+        this.units = units
+        this.allowedRegion = allowedRegion
+    
+    override function execute()
+        let spawnPoint = this.allowedRegion.randomPoint()
+        let targetID = this.units.random()
+    
+        // Spawn a unit of the selected type.
+        createUnit(
+            players[PLAYER_NEUTRAL_AGGRESSIVE],
+            targetID,
+            spawnPoint,
+            randomAngle()
+        )
+
+public class AnimalSpawnComposer implements Composer
+    protected Pool<int> units
+    protected Pool<rect> allowedRegions
+    protected region disallowedRegions

```

Comment:
This name is confusing. Using plural makes me think it's a collection of regions, but it's just one.